powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/internal/ValidationAspect.java

```diff
@@ -140,33 +146,88 @@ record -> validate(decode(record.getData()), inboundJsonSchema)));
             }
         }
 
-        Object result = pjp.proceed(proceedArgs);
-
-        if (validationNeeded && !validation.outboundSchema().isEmpty()) {
-            JsonSchema outboundJsonSchema = getJsonSchema(validation.outboundSchema(), true);
-
-            if (result instanceof APIGatewayProxyResponseEvent) {
-                APIGatewayProxyResponseEvent response = (APIGatewayProxyResponseEvent) result;
-                validate(response.getBody(), outboundJsonSchema);
-            } else if (result instanceof APIGatewayV2HTTPResponse) {
-                APIGatewayV2HTTPResponse response = (APIGatewayV2HTTPResponse) result;
-                validate(response.getBody(), outboundJsonSchema);
-            } else if (result instanceof APIGatewayV2WebSocketResponse) {
-                APIGatewayV2WebSocketResponse response = (APIGatewayV2WebSocketResponse) result;
-                validate(response.getBody(), outboundJsonSchema);
-            } else if (result instanceof ApplicationLoadBalancerResponseEvent) {
-                ApplicationLoadBalancerResponseEvent response = (ApplicationLoadBalancerResponseEvent) result;
-                validate(response.getBody(), outboundJsonSchema);
-            } else if (result instanceof KinesisAnalyticsInputPreprocessingResponse) {
-                KinesisAnalyticsInputPreprocessingResponse response =
-                        (KinesisAnalyticsInputPreprocessingResponse) result;
-                response.getRecords().forEach(record -> validate(decode(record.getData()), outboundJsonSchema));
-            } else {
-                LOG.warn("Unhandled response type {}, please use the 'envelope' parameter to specify what to validate",
-                        result.getClass().getName());
-            }
+        // don't execute the lambda if result was set by previous validation step
+        // in that case result should already hold a response with validation information
+        if (result != null) {
+        	LOG.error("Incoming API event's body failed inbound schema validation.");
         }
+        else {
+        	result = pjp.proceed(proceedArgs);
+        	
+        	if (validationNeeded && !validation.outboundSchema().isEmpty()) {
+        		JsonSchema outboundJsonSchema = getJsonSchema(validation.outboundSchema(), true);
+        		
+        		Object overridenResponse = null;
+        		if (result instanceof APIGatewayProxyResponseEvent) {
+        			APIGatewayProxyResponseEvent response = (APIGatewayProxyResponseEvent) result;
+        			overridenResponse = validateAPIGatewayProxyBody(response.getBody(), outboundJsonSchema);
```

Comment:
Question: if the response already contains some information that needs to be returned, should we actually get rid of that? For example if they add some headers, we probably want to pass them. Maybe you should pass the response as a parameter and just override the body/status code, but keep the rest.